lib/rbi/type_parser.rb

```diff
@@ -0,0 +1,320 @@
+# typed: strict
+# frozen_string_literal: true
+
+module RBI
+  class Type
+    class Error < RBI::Error; end
+
+    class << self
+      sig { params(string: String).returns(Type) }
+      def parse_string(string)
+        result = Prism.parse(string)
+        unless result.success?
+          raise Error, result.errors.map { |e| "#{e.message}." }.join(" ")
+        end
+
+        node = result.value
+        raise Error, "Expected a type expression, got `#{node.class}`" unless node.is_a?(Prism::ProgramNode)
+        raise Error, "Expected a type expression, got nothing" if node.statements.body.empty?
+        raise Error, "Expected a single type expression, got `#{node.slice}`" if node.statements.body.size > 1
+
+        node = T.must(node.statements.body.first)
+        parse_node(node)
+      end
+
+      sig { params(node: Prism::Node).returns(Type) }
+      def parse_node(node)
+        case node
+        when Prism::ConstantReadNode, Prism::ConstantPathNode
+          parse_constant(node)
+        when Prism::CallNode
+          parse_call(node)
+        when Prism::ArrayNode
+          parse_tuple(node)
+        when Prism::HashNode, Prism::KeywordHashNode
+          parse_shape(node)
+        when Prism::ParenthesesNode
+          body = node.body
+          raise Error, "Expected exactly 1 child, got 0" unless body.is_a?(Prism::StatementsNode)
+
+          children = body.body
+          raise Error, "Expected exactly 1 child, got #{children.size}" unless children.size == 1
+
+          parse_node(T.must(children.first))
+        else
+          raise Error, "Unexpected node `#{node}`"
+        end
+      end
+
+      private
+
+      sig { params(node: T.any(Prism::ConstantReadNode, Prism::ConstantPathNode)).returns(Type) }
+      def parse_constant(node)
+        case node
+        when Prism::ConstantReadNode
+          # `Foo`
+          Type.simple(node.slice)
```

Comment:
I know you had concerns about the parser potentially raising when parsing non-sensical input instead of parsing it as is and allowing another system to show errors.

For that reason, I wonder if we should use the constructors directly here, instead of using the helper methods.

Reply:
I'm still conflicted on the choice with made regarding the builder API.

Calling `new` directly is a good escape hatch but by making it private we made it impossible to use for external users. Maybe the right thing to do would be to keep the `new` public?

In the meantime you're right, I changed all the calls from the parser to use `new` directly rather than the builder interface.

Reply:
Agreed, but it is easier to make the `new` public later than to make it private later. So, I think I like how we are exposing a more minimal API for the time being.